Fix GpuSemaphore to support multiple threads per task [databricks]

revans2 opened the pull request:
This fixes #9485

revans2 commented:
build

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSemaphore.scala:
```diff
@@ -115,41 +116,118 @@ object GpuSemaphore {
+      hasSemaphore = false
+    }
+    // This is only an issue because we are on the thread that is supposedly blocked.
+    // So it is really more of a sanity test. In reality there should be no threads
+    // that are blocked, but one might have been added here when the semaphore was held
+    // and now it is being released so it will block.
```
I am not too clear on this comment. Could you provide an example where you think this could happen?

It seems the notify logic above is doing the right things. I was trying to come up with a scenario to match the concern, but I am not sure I have it.

revans2 replied:
I don't think it could happen. That is why it is an error.

tgravescs commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSemaphore.scala:
```diff
@@ -115,41 +116,118 @@ object GpuSemaphore {
+          if (done) {
+            moveToActive(t)
+          }
+          shouldBlockOnSemaphore = t == blockedThreads.peek
```
isn't this just saying this thread is at the head of the queue so if it doesn't have semaphore already this one should be blocked on it and try ot acquire it?  If they aren't the head of the queue then wait()

revans2 replied:
Yes I need to add in some comments to explain what is happening.

tgravescs commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSemaphore.scala:
```diff
@@ -115,41 +116,118 @@ object GpuSemaphore {
   }
 }
 
+private final class TaskInfo(val taskId: Long) extends Logging {
```
not required now but maybe high level description of this class would be good

tgravescs commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSemaphore.scala:
```diff
@@ -115,41 +116,118 @@ object GpuSemaphore {
+      semaphore.release(numPermits)
+      hasSemaphore = false
+    }
+    // This is only an issue because we are on the thread that is supposedly blocked.
```
should this say if we are on a thread that is supposedly blocked?

Similar to Alessandro not exactly sure when this would happen but the comment also indicates just in case

tgravescs commented:
all my minor comments about comments or name changes can be ignored in this pr to not delay 23.10.

revans2 commented:
build

revans2 commented:
@abellina and @tgravescs I have updated the PR to have more comments about what it is doing and cleaned up some of the code just a little. Please take another look.
